### PR TITLE
[5.2] Container can get contextual dependency for method by looking at declaring class

### PR DESCRIPTION
**Pull request by hannesvdvreken**
Usage:

``` php
class Foo
{
    public function bar(QueueContract $queue)
    {
        ...
    }
}

$container->when(Foo::class)->needs(QueueContract::class)->give(Queue::class);

$container->call('Foo@bar');
```

Works in controller methods as well.

### CODE REVIEW

**Comment by hannesvdvreken** [comment]
If you want me to do this PR against `master`, just ping me.

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -679,15 +686,16 @@ protected function getConcrete($abstract)
...
     /**
-     * Get the contextual concrete binding for the given abstract.
+     * Get the contextual concrete binding for the given concretion and abstract.
      *
+     * @param  string  $concrete
      * @param  string  $abstract
      * @return string|null
      */
-    protected function getContextualConcrete($abstract)
+    public function getContextualConcrete($concrete, $abstract)
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Why changing the visibility?

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: That change is breaking for classes that extend it.

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Because this method is not private.

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: It's only protected.

Comment (hannesvdvreken) [inline src/Illuminate/Container/Container.php]: True. What do you recommend?

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: Not sure if it's something we really need to worry about. It's unlikely anyone is messing with that method.

Ping @taylorotwell.

**Comment by taylorotwell** [comment]
What if you want to constructor inject a QueueInterface but then method inject a different QueueInterface implementation?

**Comment by hannesvdvreken** [comment]
That's a different story. Maybe a `$container->when('Foo@bar')->needs()->give();`?

**Comment by taylorotwell** [comment]
Yeah I think something like that should be the goal compared to the current implementation in this PR.

**Comment by hannesvdvreken** [comment]
Ok, gonna prepare a new PR for that. Master branch? Please warn me if you're DIY'ing it, so I don't do work you're doing as well.

**Pull request closed without merging**